File: packages/javascript/lib/word/randConcept.ts
import type { RandThemedWordOptions, WordDetail } from '../_types/global.js';
import { themedWord } from './wordGenerator.js';

/**
 * Ideas out of the humanities and the social world.
 *
 * `randWord({ theme: 'concept' })` with the theme already chosen; every other
 * option is the same.
 *
 * @example
 * randConcept({ language: 'ko', count: 3 }); // ['자유', '평화', '진리']
 * randConcept({ language: 'en', count: 3 }); // ['Freedom', 'Peace', 'Truth']
 */
export function randConcept(options?: RandThemedWordOptions & { output?: 'value' }): string[];
/**
 * Concept words with the language and theme each one came from.
 *
 * @example
 * randConcept({ language: 'ko', output: 'detail' });
 * // [{ word: '자유', language: 'ko', theme: 'concept' }]
 */
export function randConcept(options: RandThemedWordOptions & { output: 'detail' }): WordDetail[];
export function randConcept(options: RandThemedWordOptions = {}): string[] | WordDetail[] {
	return themedWord('concept', options);
}

File: packages/dart/lib/src/word/rand_concept.dart
import 'dart:math';

import 'package:randino/src/types.dart';
import 'package:randino/src/word/rand_word.dart';

/// Ideas out of the humanities and the social world.
///
/// [randWord] with the theme already chosen; every other parameter is the same.
/// For the detail form, pass [WordTheme.concept] to `randWordDetails` — Dart has
/// no overloads, and twenty-four more functions for it would be twenty-four too many.
///
/// ```dart
/// randConcept(language: WordLanguage.ko, count: 3); // [자유, 평화, 진리]
/// randConcept(language: WordLanguage.en, count: 3); // [Freedom, Peace, Truth]
/// ```
List<String> randConcept({
  WordLanguage? language,
  int count = 1,
  RandRealism realism = RandRealism.real,
  RandVocabulary vocabulary = RandVocabulary.full,
  int? minLength,
  int? maxLength,
  String? startsWith,
  bool unique = false,

  /// Where the randomness comes from: `Random.secure()` for a value nobody may
  /// predict, `Random(42)` for one that has to come out the same every run.
  Random? random,
}) => randWord(
  language: language,
  theme: WordTheme.concept,
  count: count,
  realism: realism,
  vocabulary: vocabulary,
  minLength: minLength,
  maxLength: maxLength,
  startsWith: startsWith,
  unique: unique,
  random: random,
);
